### PR TITLE
Add multiple valid audience support for CIBA call

### PR DESCRIPTION
**Pull request by KalanaDananjaya**
- Add support for token endpoint and ciba endpoint as valid audiences for CIBA(/oauth2/ciba) call(Acc. to spec requirement)

Issue: 
https://github.com/wso2/product-is/issues/14178

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Related PRs
Framework: https://github.com/wso2/carbon-identity-framework/pull/4127
Inbound Auth : https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/1876
Addons: https://github.com/wso2-extensions/identity-oauth-addons/pull/88

### CODE REVIEW

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 1]
```diff
@@ -228,12 +228,10 @@ private boolean validateIssuer(String issuer, String consumerKey) throws OAuthCl
 
     // "The Audience SHOULD be the URL of the Authorization Server's Token Endpoint", if a valid audience is not
     // specified.
-    private boolean validateAudience(String expectedAudience, List<String> audience) throws OAuthClientAuthnException {
+    private boolean validateAudience(ArrayList<String> expectedAudience, List<String> audience) throws OAuthClientAuthnException {
```
Comment (VivekVinushanth) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:231]: Shall we go with more generic collection: i.e. List

Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: Fixed

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 2]
```diff
@@ -228,12 +228,11 @@ private boolean validateIssuer(String issuer, String consumerKey) throws OAuthCl
...
 
-        for (String aud : audience) {
-            if (StringUtils.equals(expectedAudience, aud)) {
-                return true;
-            }
+        if (expectedAudience.contains(audience.get(0))) {
+            return true;
         }
         if (log.isDebugEnabled()) {
             log.debug("None of the audience values matched the tokenEndpoint Alias :" + expectedAudience);
```
Comment (VivekVinushanth) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:238]: Dont we need to update this log.

Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: updated

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java [hunk 3]
```diff
@@ -68,6 +68,7 @@ public PrivateKeyJWTClientAuthenticator() {
         int rejectBeforePeriod = DEFAULT_VALIDITY_PERIOD_IN_MINUTES;
         boolean preventTokenReuse = true;
         String tokenEPAlias = DEFAULT_AUDIENCE;
+        List<String> validAudiences = new ArrayList<>();
```
Comment (VivekVinushanth) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java:71]: Is the authenticator a method level when it obtain the lists of valid audiences? If not for concurrent requests will we not end up with ciba audiences for general (non-ciba) request?

Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java:71]: This is fixed with https://github.com/wso2-extensions/identity-oauth-addons/pull/88/commits/09259253f54f66af8b0b88ff2cc7feb57e770c2c

Only the first element of audience array(issuer Identifier) will be used if its a non CIBA call

**Review by VivekVinushanth — APPROVED** [review]

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 4]
```diff
@@ -91,7 +91,7 @@ public class JWTValidator {
 
     private JWTStorageManager jwtStorageManager;
 
-    public JWTValidator(boolean preventTokenReuse, String validAudience, int rejectBefore, String validIssuer,
+    public JWTValidator(boolean preventTokenReuse, List<String> validAudience, int rejectBefore, String validIssuer,
```
Comment (senthalan) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:94]: We can't change a public method signature. We need to keep the old definition as it is and define as @Deprecated and call the new method.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 5]
```diff
@@ -109,9 +109,9 @@ public JWTValidator(boolean preventTokenReuse, String validAudience, int rejectB
      *
      * @param signedJWT Validate the token
      * @return true if the jwt is valid.
-     * @throws IdentityOAuth2Exception
+     * @throws OAuthClientAuthnException
      */
-    public boolean isValidAssertion(SignedJWT signedJWT) throws OAuthClientAuthnException {
+    public boolean isValidAssertion(SignedJWT signedJWT, boolean isBackchannelCall) throws OAuthClientAuthnException {
```
Comment (senthalan) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:114]: We can't change a public method signature. We need to keep the old definition as it is and define as  deprecated and call the new method.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java (hunk 4 above)
Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: Ack. Readded the old method and deprecated it.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java (hunk 5 above)
Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:201]: Ack. Readded the old method and deprecated it.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java (hunk 3 above)
Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java:71]: Correction: By constructor, one can setup custom audience values for their JWT authenticator. In the default PkJWT authenticator, validAudiences is an empty string therefore this class level variable is not used. Instead, a class level variable is used which returns issuer identifier on a non ciba call, and all issuer identifier, ciba endpoint and token endpoint on a ciba call

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 6]
```diff
@@ -228,15 +228,14 @@ private boolean validateIssuer(String issuer, String consumerKey) throws OAuthCl
...
     // specified.
-    private boolean validateAudience(String expectedAudience, List<String> audience) throws OAuthClientAuthnException {
+    private boolean validateAudience(List<String> expectedAudience, List<String> audience)
+            throws OAuthClientAuthnException {
 
-        for (String aud : audience) {
-            if (StringUtils.equals(expectedAudience, aud)) {
-                return true;
-            }
+        if (expectedAudience.contains(audience.get(0))) {
```
[the remark this answers is not in the document]
Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: this is updated to check if any one of the given audience values are available in the valid audience values

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 7]
```diff
@@ -175,6 +191,77 @@ public boolean isValidAssertion(SignedJWT signedJWT) throws OAuthClientAuthnExce
...
+                    , timeStampSkewMillis) || !validateNotBeforeClaim(currentTimeInMillis, timeStampSkewMillis, nbf) ||
+                    !validateAgeOfTheToken(issuedAtTime, currentTimeInMillis, timeStampSkewMillis) || !isValidSignature
+                    (consumerKey, signedJWT, tenantDomain, jwtSubject)) {
+                return false;
+            }
+
+            return true;
+
+        } catch (IdentityOAuth2Exception e) {
+            return logAndThrowException(e.getMessage());
```
Comment (KalanaDananjaya) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:261]: Pass the exception

Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java]: A new exception is thrown using logAndThrowException method across the whole class

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java [hunk 8]
```diff
@@ -109,10 +109,24 @@ public JWTValidator(boolean preventTokenReuse, String validAudience, int rejectB
      *
      * @param signedJWT Validate the token
      * @return true if the jwt is valid.
-     * @throws IdentityOAuth2Exception
+     * @throws OAuthClientAuthnException
      */
+    @Deprecated
```
Comment (senthalan) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:114]: in the java doc let's point out what is the new method which needs to be used.

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java [hunk 9]
```diff
@@ -98,7 +98,11 @@ public PrivateKeyJWTClientAuthenticator() {
     public boolean authenticateClient(HttpServletRequest httpServletRequest, Map<String, List> bodyParameters,
                                       OAuthClientAuthnContext oAuthClientAuthnContext) throws OAuthClientAuthnException {
 
-        return jwtValidator.isValidAssertion(getSignedJWT(bodyParameters, oAuthClientAuthnContext));
+        if (isBackchannelCall(httpServletRequest)) {
+            return jwtValidator.isValidAssertion(getSignedJWT(bodyParameters, oAuthClientAuthnContext), true);
+        } else {
+            return jwtValidator.isValidAssertion(getSignedJWT(bodyParameters, oAuthClientAuthnContext), false);
+        }
```
Comment (senthalan) [inline component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java:105]:
```suggestion
return jwtValidator.isValidAssertion(getSignedJWT(bodyParameters, oAuthClientAuthnContext), isBackchannelCall(httpServletRequest));
```

Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/PrivateKeyJWTClientAuthenticator.java]: Fixed with https://github.com/wso2-extensions/identity-oauth-addons/pull/88/commits/2124689721ed04d73e1e579e5b12ee117d6b352c

File: component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java (hunk 8 above)
Comment (KalanaDananjaya) [reply component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java:115]: Fixed with https://github.com/wso2-extensions/identity-oauth-addons/pull/88/commits/2124689721ed04d73e1e579e5b12ee117d6b352c

**Review by senthalan — APPROVED** [review]

**Pull request merged**